hugr-passes/src/dataflow/datalog.rs

```diff
@@ -0,0 +1,346 @@
+//! [ascent] datalog implementation of analysis.
+//! Since ascent-(macro-)generated code generates a bunch of warnings,
+//! keep code in here to a minimum.
+#![allow(
+    clippy::clone_on_copy,
+    clippy::unused_enumerate_index,
+    clippy::collapsible_if
+)]
+
+use ascent::lattice::{BoundedLattice, Lattice};
+use itertools::zip_eq;
+use std::cmp::Ordering;
+use std::hash::Hash;
+use std::ops::{Index, IndexMut};
+
+use hugr_core::extension::prelude::{MakeTuple, UnpackTuple};
+use hugr_core::ops::{OpTrait, OpType};
+use hugr_core::{HugrView, IncomingPort, Node, OutgoingPort, PortIndex as _};
+
+use super::{AbstractValue, DFContext, PartialValue};
+
+type PV<V> = PartialValue<V>;
+
+#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
+pub enum IO {
+    Input,
+    Output,
+}
+
+ascent::ascent! {
+    pub(super) struct AscentProgram<V: AbstractValue, C: DFContext<V>>;
+    relation context(C);
+
+    relation node(C, Node);
+    relation in_wire(C, Node, IncomingPort);
+    relation out_wire(C, Node, OutgoingPort);
+    relation parent_of_node(C, Node, Node);
+    relation io_node(C, Node, Node, IO);
+    lattice out_wire_value(C, Node, OutgoingPort, PV<V>);
+    lattice in_wire_value(C, Node, IncomingPort, PV<V>);
+    lattice node_in_value_row(C, Node, ValueRow<V>);
```

Comment:
For an input port connected directly to an output port, this stores three copies of the value, which ascent then needs to keep in sync with lattice operations. We could be more economical like this: let a "link" refer to a group of ports that are all connected. We can then store the association of node and port to link, and the value of the link. That would reduce the number of times that values are copied.

Reply:
I think this is worth noting as a future optimization but not doing yet / until benchmarks demand, as I think the extra level of indirection will reduce clarity. But to clarify (as I haven't written the comment yet), something like...
```
relation link_values(Link, PV<V>);
relation out_wire_link(Node, OutgoingPort, Link);
relation in_wire_link(Node, OutgoingPort, Link);
relation node_in_value_row(Node, Vec<Link>);
```
the last three are fairly trivial. But then, each (say) `out_wire_value(n, OutgoingPort::from(p.index()), v) <-- .....` becomes
```
link_values(lnk, v) <-- .....,
   out_wire_link(n, OutgoingPort::from(p.index()), lnk)
```
?
That didn't seem too bad, hmmm. I can add a comment to that effect, or make a follow-up PR that I'd keep separate (in case we ever need to revert for debugging, say) if you think that's worth it?

Reply:
Ok, so I had a bit more of a go. Defining a `Link` as a `(Node, OutgoingPort)` (it'd be better to use a Wire but there), I defined:
```
relation node_in_value_preds(Node, Vec<(Node, OutgoingPort)>);
```
which could be filled straightforwardly from the edges of the graph (an element being a node, and then the source of each of said node's inputs). The best I could come up with for a standard-control-flow "untuple the first element and append the rest" was then....
```
             for (succ_n, succ) in hugr.output_neighbours(*pred).enumerate(),
             output_child(pred, out_n),
             _cfg_succ_dest(cfg, succ, dest),
-            node_in_value_row(out_n, out_in_row),
-            if let Some(fields) = out_in_row.unpack_first(succ_n, df_block.sum_rows.get(succ_n).unwrap().len()),
-            for (out_p, v) in fields.enumerate();
+            node_in_value_preds(out_n, out_in_row),
+            for (idx, (out_node, out_port)) in out_in_row.iter().enumerate(),
+            out_wire_value(out_node, out_port, v1),
+            let first_tuple_len = df_block.sum_rows.get(succ_n).unwrap().len(),
+            for (out_p, v) in if idx==0 {
+                v1.variant_values(succ_n, first_tuple_len).unwrap_or_default()
+                .into_iter().enumerate().collect()
+            } else {
+                vec![(idx + first_tuple_len, v1.clone())]
+            };
```
which....unless there's some cunning way to make datalog procedures that I'm missing, would be needed about 5 times. (Bad, but not terrible). However I got even more stuck on the call to `propagate_leaf_op` where it's not sufficient to define in-values to the leaf op one IncomingPort at a time, we need to reassemble them into a Vector to pass the whole lot together to `propagate_leaf_op`. AFAICS this would require defining an ascent custom aggregator, and at that point I'm thinking - well we could, but no, let's not do this now, and let's hope we never have to....

Reply:
There's much simpler and less cursed ways to do this; but I agree, this PR doesn't need it.